lib/Agrammon/OutputFormatter/GUI.pm6

```diff
@@ -67,6 +56,48 @@ sub get-data($model, $outputs) {
     return @records;
 }
 
+sub make-record($fq-name, $output, $model, $raw-value, $var, $filter-id?) {
+    my $format = $model.output-format($fq-name, $output);
+    my $full-value = flat-value($raw-value);
+    my $value = ($format  && $full-value.defined) ?? sprintf($format, $full-value)
+                                                  !! $full-value;
+    my $filter;
+    if $filter-id {
+        $filter-id ~~ / '=' (.+) /;
+        $filter = ~$0;
+    }
+    return %(
+        :$format,
+        :print($model.output-print($fq-name, $output)),
+        :order($model.output-order($fq-name, $output)),
+        :labels($model.output-labels($fq-name, $output)),
+        :units($model.output-units($fq-name, $output)),
+        :fullValue($full-value),
+        :$value,
+        :$var,
+        :$filter,
+    );
+}
+
+sub push-filters(@records, $fq-name, $output, $model,
+                 Agrammon::Outputs::FilterGroupCollection $collection,
+                 $var) {
+    my @results = $collection.results-by-filter-group;
+    for @results {
+        my %filters := .key;
+        my $value := .value;
+        my @filters = %filters.map: { .key ~ '=' ~ .value };
+        for @filters.kv -> $idx, $filter-id {
+            push @records, make-record($fq-name, $output, $model, $value, $var, $filter-id);
+            # TODO: what does this do in text formatter? Do we need it here?
```

Comment:
@jnthn Why is this test for `$idx` needed? Do I need something similar here?

Reply:
@zaucker Not exactly similar because you're producing a data structure rather than text, but you need to do *something*.

A filter may involve multiple key/value pairs (even though at the moment we only ever have one input with `filter = true` in a module). The loop here with the index is making sure that we only output the value and unit once (on the same line as the filter ID), and then if there are other inputs that make up the filter ID those come on the lines below. Basically, it's trying to show composite filter IDs.

You probably don't want the loop here (nor producing the `@filters` array with strings of text), but rather to do something more like:

```
push @records, make-record($fq-name, $output, $model, $value, $var, %filters);
```

Since at the moment the code is producing a string, only to throw a regex at it and parse it again. More deeply, though, it seems that the code in `make-record` only expects there to be one filter, and it only passes along the value. If you want to just do that, then `%filters.values[0]` would give you what you're current sending as `$filter`. Really, depends if you want to do the general thing now (IMHO, I'd just pass the `%filters` hash along to the UI as is and then deal with how to render filters in the frontend).

Reply:
Ok, thanks. I'll adapt the frontend to handle the filters.

Reply:
Fixed